### PR TITLE
oneDNN NHWC fixes

### PR DESCRIPTION
**Pull request by jczaja**
### PR types
Bug fixes 

### PR changes
Others 

### Describe
oneDNN NHWC implementation need to know upfront if model is NHWC or NCHW so mechanics was added to check that.
Also relevant UT was added from https://github.com/PaddlePaddle/Paddle/pull/39654

### CODE REVIEW

**Comment by jczaja** [comment]
@baoachun HI, This PR fixes #38126 (pointed UTs and model) and incorporates UT you have  created. The only problem is that in CI. test_lrn_op is failing (time out) . test_lrn_op is no related to oneDNN so perhaps you can advice where problem is?

**Comment by lidanqing-vv** [comment]
@baoachun Hi, this PR passed all CIs please review.

**Comment by jczaja** [comment]
@tsocha , @Silv3S Please help with review. Ask questions if something is not clear

File: paddle/fluid/platform/mkldnn_helper.h [hunk 1]
```diff
@@ -559,6 +559,34 @@ inline void GetGroupConvWeightsTz(std::vector<int64_t>& weights_tz,  // NOLINT
   }
 }
 
+inline void RegisterModelLayout(
+    std::vector<std::unique_ptr<framework::OperatorBase>>& ops,
+    const platform::Place& place) {
+  if (platform::is_cpu_place(place)) {
+    auto check_attrib = [&](std::unique_ptr<framework::OperatorBase>& op,
+                            const std::string& attrib_name) -> bool {
```
Comment (tsocha) [inline paddle/fluid/platform/mkldnn_helper.h]:
```suggestion
    auto check_attrib = [](std::unique_ptr<framework::OperatorBase>& op,
                           const std::string& attrib_name) -> bool {
```

File: python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_conv_elementwise_add_fuse_pass.py [hunk 2]
```diff
@@ -25,17 +25,120 @@
 import hypothesis.strategies as st
 
 
-class TestConvElementwiseAddMkldnnFusePass(PassAutoScanTest):
+# the two inputs of elementwise_add are tensor
+class TestConvElementwiseAddMkldnnFusePass1(PassAutoScanTest):
```
Comment (tsocha) [inline python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_conv_elementwise_add_fuse_pass.py:29]:
```suggestion
class TestConvElementwiseAddMkldnnFusePass(PassAutoScanTest):
```

File: paddle/fluid/operators/prelu_op.cc [hunk 3]
```diff
@@ -212,6 +230,24 @@ class PReluGradOp : public framework::OperatorWithKernel {
...
+        paddle::platform::MKLDNNDeviceContext::tls()
+                .get_cur_paddle_data_layout() == framework::DataLayout::kNHWC) {
+      return framework::OpKernelType(expected_kernel_type.data_type_,
+                                     tensor.place(),
+                                     framework::DataLayout::kNHWC);
+    }
+#endif
+    return framework::OpKernelType(expected_kernel_type.data_type_,
+                                   tensor.place(), tensor.layout());
+  }
```
Comment (tsocha) [inline paddle/fluid/operators/prelu_op.cc]: Don't repeat yourself, please.

File: paddle/fluid/operators/prelu_op.cc [hunk 4]
```diff
@@ -128,6 +128,24 @@ class PReluOp : public framework::OperatorWithKernel {
...
+
+  framework::OpKernelType GetKernelTypeForVar(
+      const std::string &var_name, const Tensor &tensor,
+      const framework::OpKernelType &expected_kernel_type) const {
+#ifdef PADDLE_WITH_MKLDNN
+    // All inputs (including alpha) need shape rotating
+    if ((expected_kernel_type.data_layout_ == framework::DataLayout::kMKLDNN) &&
+        (tensor.layout() != framework::DataLayout::kMKLDNN) &&
+        paddle::platform::MKLDNNDeviceContext::tls()
+                .get_cur_paddle_data_layout() == framework::DataLayout::kNHWC) {
```
Comment (tsocha) [inline paddle/fluid/operators/prelu_op.cc]: Make this if more readable like this, please.
```suggestion
    auto isKernelTypeIsONEDNN = expected_kernel_type.data_layout_ == framework::DataLayout::kMKLDNN
    if (isKernelTypeIsONEDNN  &&
        (tensor.layout() != framework::DataLayout::kMKLDNN) &&
        paddle::platform::MKLDNNDeviceContext::tls()
                .get_cur_paddle_data_layout() == framework::DataLayout::kNHWC) {
```

**Review by tsocha — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_conv_elementwise_add_fuse_pass.py (hunk 2 above)
Comment (jczaja) [reply python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_conv_elementwise_add_fuse_pass.py:29]: This UT was added from https://github.com/PaddlePaddle/Paddle/pull/39654 . So I prefer any non-critical changes to left to original author.

File: paddle/fluid/operators/prelu_op.cc (hunk 3 above)
Comment (jczaja) [reply paddle/fluid/operators/prelu_op.cc]: ok

File: paddle/fluid/operators/prelu_op.cc (hunk 4 above)
Comment (jczaja) [reply paddle/fluid/operators/prelu_op.cc]: ok

File: python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_conv_elementwise_add_fuse_pass.py [hunk 5]
```diff
@@ -25,17 +25,120 @@
...
     def is_program_valid(self, program_config: ProgramConfig) -> bool:
         attrs = [
             program_config.ops[i].attrs
             for i in range(len(program_config.ops))
         ]
-        # If the problem has been fixed, the judgment
```
Comment (Silv3S) [inline python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_conv_elementwise_add_fuse_pass.py:34]: Similar condition is in _test_mkldnn_conv_mish_fuse_pass.py_ and _test_mkldnn_depthwise_conv.py_. Should we enable testing NHWC there too?

**Review by lidanqing-vv — DISMISSED** [review]
LGTM

**Review by tsocha — DISMISSED** [review]

File: python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_conv_elementwise_add_fuse_pass.py [hunk 6]
```diff
@@ -25,17 +25,120 @@
...
+    def sample_predictor_configs(self, program_config):
+        config = self.create_inference_config(use_mkldnn=True)
+        yield config, ["relu", "conv2d", "conv2d"], (1e-5, 1e-5)
+
+    def test(self):
+        self.run_and_statis(
+            quant=False, passes=["conv_elementwise_add_mkldnn_fuse_pass"])
+
+
+'''
```
Comment (Silv3S) [inline python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_conv_elementwise_add_fuse_pass.py:128]: Why is whole original unit test commented?

File: python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_conv_elementwise_add_fuse_pass.py (hunk 5 above)
Comment (jczaja) [reply python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_conv_elementwise_add_fuse_pass.py:34]: @Silv3S This UT was added in #39654 So I just copied it here as suggessted by @lidanqing-intel . I do not intend to extend it in this PR.

File: python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_conv_elementwise_add_fuse_pass.py (hunk 6 above)
Comment (jczaja) [reply python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_conv_elementwise_add_fuse_pass.py:128]: I just copied this UT from #39654 as requested so not sure of authors intention

**Review by tsocha — APPROVED** [review]

**Review by jakpiase — APPROVED** [review]
LGTM

**Comment by Silv3S** [comment]
LGTM

**Comment by jczaja** [comment]
@baoachun Please review.

**Review by lidanqing-vv — APPROVED** [review]
LGTM

**Pull request merged**